File: files/400-499/454/Challenge 454.py
import re
import pandas as pd

path = "400-499/454/CH-454 Replacement.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=2, nrows=6)
test = pd.read_excel(path, usecols="F:H", skiprows=2, nrows=6)
test.columns = input.columns


def replace_x(product):
    return re.sub(
        r"XXXX|XX", lambda match: "XX" if len(match[0]) == 4 else "X", product
    )


result = input.assign(**{"Product ID": input["Product ID"].map(replace_x)})
print(result.equals(test))

File: files/400-499/454/Challenge 454.R
library(tidyverse)
library(readxl)

path <- "400-499/454/CH-454 Replacement.xlsx"
input <- read_excel(path, range = "B3:D9")
test <- read_excel(path, range = "F3:H9")

replace_x <- function(product) {
  str_replace_all(product, "XXXX|XX", \(match) {
    if_else(str_length(match) == 4, "XX", "X")
  })
}

result <- input %>%
  mutate(`Product ID` = map_chr(`Product ID`, replace_x)) %>%
  select(all_of(names(test)))

all.equal(result, test)
